### PR TITLE
[FG:InPlacePodVerticalScaling] Gate Disallow in-place resize for guaranteed pods on nodes with a static topology policy

### LINKED ISSUE #128068: [FG:InPlacePodVerticalScaling] Disable in-place resize for guaranteed pods on nodes with a static topology policy
See https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#static-cpu--memory-policy

We probably want to introduce a new feature gate to allow resize in this case to unblock development on https://github.com/kubernetes/kubernetes/issues/127262.

/sig node
/kind feature
/priority important-longterm

### PR DESCRIPTION
**Pull request by esotsal**
#### What this PR does / why we need it:

We  want to introduce a new feature gate to allow resize  for guaranteed pods with integer CPU requests on nodes with a static topology policy to unblock development on https://github.com/kubernetes/kubernetes/issues/127262.

New gate "InPlacePodVerticalScalingExclusiveCPUs" is off by default, but can be enabled to unblock development of Static CPU management alongside InPlacePodVerticalScaling.

#### Which issue(s) this PR fixes:
Update: Attempt to fix #128068 

#### Does this PR introduce a user-facing change?

```release-note
- Ensure resizing for Guaranteed pods with integer CPU requests on nodes with static CPU & Memory policy configured is not allowed for the beta release of in-place resize. The feature gate `InPlacePodVerticalScalingExclusiveCPUs` defaults to `false`, but can be enabled to unblock development on ([#127262](https://github.com/kubernetes/kubernetes/issues/127262), [@tallclair](https://github.com/tallclair)) [SIG Node].
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
? Please advise
```
/sig node
/priority important-soon
/milestone v1.32

### CODE REVIEW

**Review by tallclair — COMMENTED** [review]
Thanks for taking this on. I see this is tagged "WIP" so I assume it's still coming, but the feature gate should be implemented in this same PR (I mean rejecting the resize of exclusive CPU pods)

File: pkg/features/versioned_kube_features.go [hunk 1]
```diff
@@ -39,6 +39,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 		{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Deprecated},
 	},
 
+	AllowResourceResizeInContainerWithExclusiveCPUs: {
+		{Version: version.MustParse("1.32"), Default: false, PreRelease: featuregate.Deprecated},
```
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/versioned_kube_features.go:43]: Thought Deprecated , because it will be incomplete ie when set to true it will be silently ignored since the functionality will not be there yet until fix is merged. I don’t have a strong opinion on that, I am ok with alpha as well if that is acceptable.

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -42,6 +42,14 @@ const (
 	// with DNS names.
 	AllowDNSOnlyNodeCSR featuregate.Feature = "AllowDNSOnlyNodeCSR"
 
+	// owner: @tallclair @esotsal
+	//
+	// Allow resource resize for containers in Guaranteed pods with integer CPU requests ( default false ).
+	// Applies only in nodes with InPlacePodVerticalScaling and CPU Manager features enabled, and
+	// CPU Manager Static Policy option set.
+	// Remove after InPlacePodVerticalScaling is GA based on user feedback that should be reflected in #1287 update.
+	AllowResourceResizeInContainerWithExclusiveCPUs featuregate.Feature = "AllowResourceResizeInContainerWithExclusiveCPUs"
```
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/kube_features.go:51]: Ack

File: pkg/features/kube_features.go [hunk 3]
```diff
@@ -42,6 +42,14 @@ const (
 	// with DNS names.
 	AllowDNSOnlyNodeCSR featuregate.Feature = "AllowDNSOnlyNodeCSR"
 
+	// owner: @tallclair @esotsal
```
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/kube_features.go:45]: Ack

File: CHANGELOG/CHANGELOG-1.32.md [hunk 4]
```diff
@@ -165,6 +165,7 @@ name | architectures
 
 ### Other (Cleanup or Flake)
 
+- Ensure resizing for Guaranteed pods on nodes with static CPU & Memory policy configured is not allowed for the beta release of in-place resize. The feature gate `AllowResourceResizeInContainerWithExclusiveCPUs` defaults to `false`, but can be enabled to unblock development on ([#127262](https://github.com/kubernetes/kubernetes/issues/127262), [@tallclair](https://github.com/tallclair)) [SIG Node].
```
[the remark this answers is not in the document]
Comment (esotsal) [reply CHANGELOG/CHANGELOG-1.32.md:168]: Ack

**Comment by esotsal** [comment]
> Thanks for taking this on. I see this is tagged "WIP" so I assume it's still coming, but the feature gate should be implemented in this same PR (I mean rejecting the resize of exclusive CPU pods)

Correct , that’s the plan, plus necessary tests

File: CHANGELOG/CHANGELOG-1.32.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply CHANGELOG/CHANGELOG-1.32.md]: Done

File: pkg/features/kube_features.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/kube_features.go:45]: Done

File: pkg/features/versioned_kube_features.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/versioned_kube_features.go]: Sorry missed comment in Slack, thanks _"This shouldn't be deprecated, since we're not removing the functionality. We're just holding back a subset of the functionality in alpha."._ Done

File: pkg/features/kube_features.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply pkg/features/kube_features.go]: Done

**Comment by esotsal** [comment]
> > Thanks for taking this on. I see this is tagged "WIP" so I assume it's still coming, but the feature gate should be implemented in this same PR (I mean rejecting the resize of exclusive CPU pods)
> 
> Correct , that’s the plan, plus necessary tests

@tallclair  PTAL

File: test/e2e_node/pod_resize_test.go
Comment (AnishShah) [inline test/e2e_node/pod_resize_test.go:1]: Would it be possible to avoid duplicate code from [test/e2e/common/node/pod_resize.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/node/pod_resize.go)?

To avoid duplicated code, I put the pod_resize tests in a common location [test/e2e/common/node/pod_resize.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/node/pod_resize.go) (in #128143) so they could be run as both e2e_node and cluster e2e tests.

However, your changes require modifying the kubelet configuration, which is only possible in `e2e_node` tests.

To solve this, should we:

1. Should we move (instead of copy) the pod_resize tests from `test/e2e/common/node` to `test/e2e_node` ? This way we would avoid duplicate code.
2. If we do that, will we lose any test coverage by not running these tests as cluster e2e tests?

what do you think? 

cc @tallclair, do you have any suggestions as well?

File: test/e2e_node/pod_resize_test.go [hunk 5]
```diff
@@ -0,0 +1,1768 @@
...
+			framework.ExpectNoError(err)
+			// Resize has been actuated, test rollback
+			patchAndVerify(rbPatchStr, tc.containers, tc.expected, "rollback", true)
+
+			ginkgo.By("deleting pod")
+			deletePodSyncByName(ctx, f, newPod.Name)
+			// we need to wait for all containers to really be gone so cpumanager reconcile loop will not rewrite the cpu_manager_state.
+			// this is in turn needed because we will have an unavoidable (in the current framework) race with the
+			// reconcile loop which will make our attempt to delete the state file and to restore the old config go haywire
+			waitForAllContainerRemoval(ctx, newPod.Name, newPod.Namespace)
```
Comment (AnishShah) [inline test/e2e_node/pod_resize_test.go:1449]: `deletePodSyncByName` will wait until the pod is delete. Do we need `waitForAllContainerRemoval` in that case? I would expect that the pod containers are really gone by the time pod is deleted.

File: test/e2e_node/pod_resize_test.go [hunk 6]
```diff
@@ -0,0 +1,1768 @@
...
+			ginkgo.By("verifying pod allocations after patch")
+			gomega.Eventually(ctx, e2epod.VerifyPodAllocations, timeouts.PodStartShort, timeouts.Poll).
+				WithArguments(patchedPod, tc.expected).
+				Should(gomega.BeNil(), "failed to verify Pod allocations for patchedPod")
+
+			deletePodSyncByName(ctx, f, newPod.Name)
+			// we need to wait for all containers to really be gone so cpumanager reconcile loop will not rewrite the cpu_manager_state.
+			// this is in turn needed because we will have an unavoidable (in the current framework) race with the
+			// reconcile loop which will make our attempt to delete the state file and to restore the old config go haywire
+			waitForAllContainerRemoval(ctx, newPod.Name, newPod.Namespace)
```
Comment (AnishShah) [inline test/e2e_node/pod_resize_test.go:1578]: same comment here. Do we need `waitForAllContainerRemoval`?

File: test/e2e_node/pod_resize_test.go (hunk 5 above)
Comment (esotsal) [reply test/e2e_node/pod_resize_test.go:1449]: Did thorough tests, yest we need it. I was getting flaky tests.

File: test/e2e_node/pod_resize_test.go (hunk 6 above)
Comment (esotsal) [reply test/e2e_node/pod_resize_test.go:1578]: Did thorough tests, yest we need it. I was getting flaky tests.

File: test/e2e/framework/pod/resize.go [hunk 7]
```diff
@@ -72,12 +73,13 @@ type ContainerAllocations struct {
...
-	CPUPolicy    *v1.ResourceResizeRestartPolicy
-	MemPolicy    *v1.ResourceResizeRestartPolicy
-	RestartCount int32
+	Name                 string
+	Resources            *ContainerResources
+	Allocations          *ContainerAllocations
+	CPUPolicy            *v1.ResourceResizeRestartPolicy
+	MemPolicy            *v1.ResourceResizeRestartPolicy
+	RestartCount         int32
+	CpusAllowedListValue string
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go:82]: Ack

File: test/e2e/framework/pod/resize.go [hunk 8]
```diff
@@ -456,3 +458,58 @@ func ResizeContainerPatch(containers []ResizableContainerInfo) (string, error) {
 
 	return string(patchBytes), nil
 }
+
+// TODO TO replace with cpuset method
+// Credits github.com/prometheus/procfs/proc_status.go
+func calcCpusAllowedList(cpuString string) []uint64 {
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go:464]: Ack

File: test/e2e/framework/pod/resize.go [hunk 9]
```diff
@@ -456,3 +458,58 @@ func ResizeContainerPatch(containers []ResizableContainerInfo) (string, error) {
 
 	return string(patchBytes), nil
 }
+
+// TODO TO replace with cpuset method
+// Credits github.com/prometheus/procfs/proc_status.go
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go:463]: Ack

File: test/e2e/framework/pod/resize.go [hunk 10]
```diff
@@ -456,3 +458,58 @@ func ResizeContainerPatch(containers []ResizableContainerInfo) (string, error) {
...
+			g = append(g, cpu)
+		}
+
+	}
+
+	sort.Slice(g, func(i, j int) bool { return g[i] < g[j] })
+	return g
+}
+
+func VerifyPodContainersCpusAllowedListValue(f *framework.Framework, pod *v1.Pod, wantCtrs []ResizableContainerInfo) error {
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go:489]: Ack

File: test/e2e/common/node/pod_resize.go [hunk 11]
```diff
@@ -887,6 +888,11 @@ func doPodResizeTests(f *framework.Framework) {
 			ginkgo.By("verifying initial pod resize policy is as expected")
 			e2epod.VerifyPodResizePolicy(newPod, tc.containers)
 
+			// Pod must be both running and must have ready condition of status true, below are needed to solve Flaky tests, CreateSync ensures only that pod is Running.
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/common/node/pod_resize.go:891]: That's true, 
https://github.com/kubernetes/kubernetes/blob/74b9204b6a6bd5cce1c8b7c68bd2f39caf1b7fcb/pkg/api/pod/util.go#L250-L255

but it is not checking PodRunningReady condition , this is why i have added it later.

https://github.com/kubernetes/kubernetes/blob/74b9204b6a6bd5cce1c8b7c68bd2f39caf1b7fcb/test/utils/conditions.go#L31-L45

[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/common/node/pod_resize.go:891]: I was getting flaky tests without it, this is why i added it.

File: test/e2e/framework/node/helper.go [hunk 12]
```diff
@@ -175,3 +178,22 @@ func IsARM64(node *v1.Node) bool {
...
+	} else {
+		re := regexp.MustCompile("containerd://(.*)")
+		match := re.FindStringSubmatch(node.Status.NodeInfo.ContainerRuntimeVersion)
+		if len(match) != 2 {
+			return false
+		}
+		// TODO(InPlacePodVerticalScaling): Update when RuntimeHandlerFeature for pod resize have been implemented
+		if ver, verr := semver.ParseTolerant(match[1]); verr == nil {
+			return ver.Compare(semver.MustParse(e2epod.MinContainerRuntimeVersion)) >= 0
+		}
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/node/helper.go:196]: Ack

File: test/e2e_node/pod_resize_test.go
Comment (esotsal) [reply test/e2e_node/pod_resize_test.go:1]: > * Should we move (instead of copy) the pod_resize tests from `test/e2e/common/node` to `test/e2e_node` ? This way we would avoid duplicate code.

I believe so.

> 
> * If we do that, will we lose any test coverage by not running these tests as cluster e2e tests?

Checking the runs, seems tests today assume one node anyhow, so i believe we don't lose test coverage. Without e2e_node we certainly lose test coverage.

File: test/e2e/framework/pod/resize.go [hunk 13]
```diff
@@ -456,3 +458,58 @@ func ResizeContainerPatch(containers []ResizableContainerInfo) (string, error) {
...
+// TODO TO replace with cpuset method
+// Credits github.com/prometheus/procfs/proc_status.go
+func calcCpusAllowedList(cpuString string) []uint64 {
+	s := strings.Split(cpuString, ",")
+
+	var g []uint64
+
+	for _, cpu := range s {
+		// parse cpu ranges, example: 1-3=[1,2,3]
+		if l := strings.Split(strings.TrimSpace(cpu), "-"); len(l) > 1 {
```
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go:471]: Ack, will check and comeback.

File: test/e2e_node/pod_resize_test.go
Comment (esotsal) [reply test/e2e_node/pod_resize_test.go:1]: I think better test coverage is more important than having some duplicate code. I will try to reduce double parts in the meantime to just enough.

File: test/e2e/framework/node/helper.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/node/helper.go]: Done

File: test/e2e/framework/pod/resize.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go]: Done

File: test/e2e/framework/pod/resize.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go]: Done

File: test/e2e/common/node/pod_resize.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tallclair) [reply test/e2e/common/node/pod_resize.go:891]: It's the same condition that CreateSync is checking. Either there's a bug in one of these checks, or there's something else going on. I'm betting on the latter. Maybe you just got lucky and the test didn't flake after adding this?

File: test/e2e_node/pod_resize_test.go
Comment (tallclair) [reply test/e2e_node/pod_resize_test.go:1]: I think it's important that we run the resize tests in cluster E2E, although I could see an argument for running a much smaller subset of them.

As it stands, this test has far too much duplicate code and is unmaintainable.

IMO, we should have a separate e2e node test job with static cpu policy enabled. Then we don't need to worry about changing out the kubelet config in these tests, and can just share the common tests.

File: pkg/features/kube_features.go [hunk 14]
```diff
@@ -739,6 +739,15 @@ const (
 	// Enables In-Place Pod Vertical Scaling
 	InPlacePodVerticalScaling featuregate.Feature = "InPlacePodVerticalScaling"
 
+	// owner: @tallclair @esotsal
+	// alpha: v1.32
+	//
+	// Allow resource resize for containers in Guaranteed pods with integer CPU requests ( default false ).
+	// Applies only in nodes with InPlacePodVerticalScaling and CPU Manager features enabled, and
+	// CPU Manager Static Policy option set.
+	// Remove after InPlacePodVerticalScaling is GA based on user feedback that should be reflected in #1287 update.
```
Comment (tallclair) [inline pkg/features/kube_features.go:748]: I think delete this line. Removal after GA with sufficient user feedback is assumed for all feature gate removal.

```suggestion
```

File: test/e2e/common/node/pod_resize.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tallclair) [reply test/e2e/common/node/pod_resize.go:891]: BTW, I also was hitting a bunch of flakes in these tests, and I'm working on improving the situation in https://github.com/kubernetes/kubernetes/pull/128377. There's a lot more going there than just waiting for the pod to be ready.

**Review by tallclair — COMMENTED** [review]
I don't think this is fully disabling resize of pods using exclusive CPUs.

If the feature gate is disabled but a pod is using guaranteed CPUs, then any resize should be marked as infeasible. I think the place to implement that is in handlePodResourcesResize here: https://github.com/kubernetes/kubernetes/blob/ca797af2afa52c43c6139ed7f994635a94005300/pkg/kubelet/kubelet.go#L2813

If the allocation is updated, that means a resize occurred. If it's a guaranteed pod, and the node is configured with static CPU or memory policy, and the fetaure gate is disabled, then set the resize status to infeasible and return the `allocatedPod`.

I think you can leave the E2E tests off of this PR, and add them separately.

**Comment by drewhagen** [comment]
👋 @AnishShah @tallclair @kow3ns @liggitt @esotsal
I see this PR is quite active lately and tagged to go with the `1.32` release.

Friendly reminder that the code freeze is starting `02:00 UTC Friday November 8th 2024` (a little less than **1 week from now**). Please make sure this PR has both `lgtm` and `approved` labels before the code freeze. Thanks! 🚀

**Comment by esotsal** [comment]
> I don't think this is fully disabling resize of pods using exclusive CPUs.
> 
> If the feature gate is disabled but a pod is using guaranteed CPUs, then any resize should be marked as infeasible. I think the place to implement that is in handlePodResourcesResize here:
> 
> https://github.com/kubernetes/kubernetes/blob/ca797af2afa52c43c6139ed7f994635a94005300/pkg/kubelet/kubelet.go#L2813
> 
> If the allocation is updated, that means a resize occurred. If it's a guaranteed pod, and the node is configured with static CPU or memory policy, and the fetaure gate is disabled, then set the resize status to infeasible and return the `allocatedPod`.

Ack
 
> I think you can leave the E2E tests off of this PR, and add them separately.

Ack

File: pkg/features/kube_features.go (hunk 14 above)
Comment (esotsal) [reply pkg/features/kube_features.go]: Done

File: test/e2e/common/node/pod_resize.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/common/node/pod_resize.go]: Done, removed the extra check.

File: test/e2e_node/pod_resize_test.go
Comment (esotsal) [reply test/e2e_node/pod_resize_test.go:1]: > IMO, we should have a separate e2e node test job with static cpu policy enabled. Then we don't need to worry about changing out the kubelet config in these tests, and can just share the common tests.

Sounds reasonable but, please note we will need 22 separate jobs to capture all possible policy static combinations ( General availability (2) , Beta (1), Alpha (8) = 11 x 2 to cover InPlacePodVerticalScalingExclusiveCPUs ).

IMO, tests should be common i.e. not separate tests for policy static configuration ( if needed control flow depending on enabled features gates should be added )

If the 22 jobs are considered unmaintainable then plan b could be to add an e2e/node kubelet config change method ( using scp to trasnfer modified kubelet config and perform a kubelet restart as done ? )

File: test/e2e/framework/pod/resize.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go]: Refactored, now using [cpuset.Parse](https://github.com/kubernetes/utils/blob/49e7df575cb6/cpuset/cpuset.go#L203) and [cpuset.Size()](https://github.com/kubernetes/utils/blob/49e7df575cb6/cpuset/cpuset.go#L63) instead.

File: test/e2e/framework/pod/resize.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/framework/pod/resize.go]: Refactored, using methods [cpuset.Parse](https://github.com/kubernetes/utils/blob/49e7df575cb6/cpuset/cpuset.go#L203) and [cpuset.Size](https://github.com/kubernetes/utils/blob/49e7df575cb6/cpuset/cpuset.go#L63) instead.

File: test/e2e/common/node/pod_resize.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (esotsal) [reply test/e2e/common/node/pod_resize.go]: @AnishShah and @tallclair you where correct. It was luck :-) tests passed removing the extra check. As you can see it fails in two last tests which i believe has something to do with the way i use the test framework so i guess a timeout in the last run ( tries to restart kubelet but end tests asynchronously ?) . If you have time would appreciate to take a look on the two failing last tests ( it is actually one per job since two runs are performed) it has been bothering me the last week.

**Comment by esotsal** [comment]
> > I don't think this is fully disabling resize of pods using exclusive CPUs.
> > If the feature gate is disabled but a pod is using guaranteed CPUs, then any resize should be marked as infeasible. I think the place to implement that is in handlePodResourcesResize here:
> > https://github.com/kubernetes/kubernetes/blob/ca797af2afa52c43c6139ed7f994635a94005300/pkg/kubelet/kubelet.go#L2813
> > 
> > If the allocation is updated, that means a resize occurred. If it's a guaranteed pod, and the node is configured with static CPU or memory policy, and the fetaure gate is disabled, then set the resize status to infeasible and return the `allocatedPod`.
> 
> Ack
> 
 @tallclair PTAL

**Comment by esotsal** [comment]
@AnishShah removed the e2e_node tests, will update #123319  later with the latest version including resolved  comments.

**Comment by esotsal** [comment]
@tallclair @AnishShah  added unit tests can you please take a look ?

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 15]
```diff
@@ -369,9 +369,22 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+				// TODO Fix InPlacePodVerticalScaling alognside CPU Static policy, until fixed skipping pod resize
+				p.updateCPUsToReuse(pod, container, cpuset)
+				klog.InfoS("Static policy: InPlacePodVerticalScaling alognside CPU Static policy not implemented, skipping pod resize")
+				return nil
+			} else {
+				p.updateCPUsToReuse(pod, container, cpuset)
+				klog.InfoS("Static policy: InPlacePodVerticalScaling alognside CPU Static policy requires InPlacePodVerticalScaling to be enabled, skipping pod resize")
+				return nil
+			}
```
Comment (tallclair) [inline pkg/kubelet/cm/cpumanager/policy_static.go:382]: what's the difference between these 2 cases?

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 16]
```diff
@@ -369,9 +369,22 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		}
 	}
 	if cpuset, ok := s.GetCPUSet(string(pod.UID), container.Name); ok {
-		p.updateCPUsToReuse(pod, container, cpuset)
-		klog.InfoS("Static policy: container already present in state, skipping", "pod", klog.KObj(pod), "containerName", container.Name)
-		return nil
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScalingExclusiveCPUs) {
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```
Comment (tallclair) [inline pkg/kubelet/cm/cpumanager/policy_static.go:373]: Is this only called on resize? Or does this also get executed when bringing up a new pod?

File: pkg/kubelet/cm/cpumanager/policy_static_test.go
Comment (tallclair) [inline pkg/kubelet/cm/cpumanager/policy_static_test.go:1]: Revert these changes?

File: pkg/kubelet/kubelet.go [hunk 17]
```diff
@@ -2765,6 +2765,21 @@ func isPodResizeInProgress(pod *v1.Pod, podStatus *kubecontainer.PodStatus) bool
...
+	if v1qos.GetPodQOS(pod) == v1.PodQOSGuaranteed {
+		if utilfeature.DefaultFeatureGate.Enabled(features.CPUManager) {
+			if kl.containerManager.GetNodeConfig().CPUManagerPolicy == "static" {
+				klog.V(3).InfoS("Resize is infeasible for Guaranteed Pods alongside CPU Manager static policy")
+				return false, v1.PodResizeStatusInfeasible
+			}
+		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.MemoryManager) {
+			if kl.containerManager.GetNodeConfig().ExperimentalMemoryManagerPolicy == "static" {
+				klog.V(3).InfoS("Resize is infeasible for Guaranteed Pods alognside Memory Manager static policy")
```
Comment (tallclair) [inline pkg/kubelet/kubelet.go:2777]:
```suggestion
				klog.V(3).InfoS("Resize is infeasible for Guaranteed Pods alongside Memory Manager static policy")
```

File: pkg/kubelet/kubelet.go [hunk 18]
```diff
@@ -2765,6 +2765,21 @@ func isPodResizeInProgress(pod *v1.Pod, podStatus *kubecontainer.PodStatus) bool
 // pod should hold the desired (pre-allocated) spec.
 // Returns true if the resize can proceed.
 func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, v1.PodResizeStatus) {
+	if v1qos.GetPodQOS(pod) == v1.PodQOSGuaranteed {
```
Comment (tallclair) [inline pkg/kubelet/kubelet.go:2768]:
```suggestion
	if v1qos.GetPodQOS(pod) == v1.PodQOSGuaranteed  && !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScalingExclusiveCPUs) {
```

File: pkg/kubelet/cm/cpumanager/policy_static_test.go
Comment (esotsal) [reply pkg/kubelet/cm/cpumanager/policy_static_test.go:1]: I have added missing unit tests for InPlacePodVerticalScaling along side CPU Manager.

File: pkg/kubelet/kubelet.go (hunk 18 above)
Comment (esotsal) [reply pkg/kubelet/kubelet.go]: Thanks, Done

File: pkg/kubelet/kubelet.go (hunk 17 above)
Comment (esotsal) [reply pkg/kubelet/kubelet.go]: Thanks, Done

File: pkg/kubelet/cm/cpumanager/policy_static_test.go
Comment (esotsal) [reply pkg/kubelet/cm/cpumanager/policy_static_test.go:1]: Thinking it twice, i understand however your reasoning that since we don't update policy_static.go in this PR we should not update the test. Since tests have shown that it works, i remove it from the commit and we can take it later.

**Comment by wendy-ha18** [comment]
Hi @pacoxu @tallclair 
Appreciate all of your efforts with this PR!!
A gentle reminder that the code freeze has started 02:00 UTC Friday November 8th 2024 . Please make sure this has both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md).

**Comment by sftim** [comment]
Make sure to document `InPlacePodVerticalScalingExclusiveCPUs` for v1.32
(it's OK to file a k/website issue and delegate; in fact, I recommend it)

**Comment by esotsal** [comment]
> Make sure to document `InPlacePodVerticalScalingExclusiveCPUs` for v1.32 (it's OK to file a k/website issue and delegate; in fact, I recommend it)

https://github.com/kubernetes/website/issues/48687 created @sftim , thanks !

**Comment by esotsal** [comment]
Adding for tracking purposes, this PR is related with https://github.com/kubernetes/enhancements/issues/5294

**Comment by esotsal** [comment]
Adding umbrella issue for logistics 

https://github.com/kubernetes/kubernetes/issues/140749
